stan/math/prim/prob/wiener4_lccdf.hpp

```diff
@@ -0,0 +1,391 @@
+ * @tparam T_v type of drift rate
+ *
+ * @param a The boundary separation
+ * @param w_value The relative starting point
+ * @param v_value The drift rate
+ * @return log probability to reach the upper bound
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto wiener_prob(const T_a& a, const T_v& v_value, const T_w& w_value) {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  const auto v = -v_value;
+  const auto w = 1 - w_value;
+  if (fabs(v) == 0.0) {
+    return ret_t(log1p(-w));
+  }
+
+  const auto exponent = -2.0 * v * a * (1.0 - w);
+  if (exponent < 0) {
+    return ret_t(log1m_exp(exponent) - log_diff_exp(2 * v * a * w, exponent));
+  } else {
+    return ret_t(log1m_exp(-exponent) - log1m_exp(2 * v * a));
+  }
+}
+
+/**
+ * Calculate parts of the partial derivatives for wiener_prob_grad_a and
+ * wiener_prob_grad_v (on log-scale)
+ *
+ * @tparam T_a type of boundary
+ * @tparam T_w type of relative starting point
+ * @tparam T_v type of drift rate
+ *
+ * @param a The boundary separation
+ * @param w_value The relative starting point
+ * @param v_value The drift rate
+ * @return 'ans' term
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto wiener_prob_derivative_term(const T_a& a, const T_v& v_value,
+                                        const T_w& w_value) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  const auto exponent_m1 = log1p(-1.1 * 1.0e-8);
+  ret_t ans;
+  const auto v = -v_value;
+  const auto w = 1 - w_value;
+
+  if (v < 0) {
+    const auto exponent_with_1mw = 2.0 * v * a * (1.0 - w);
+    const auto exponent_with_w = 2 * a * v * w;
+    const auto exponent = 2 * a * v;
+
+    if (((exponent_with_1mw >= exponent_m1) || (exponent_with_w >= exponent_m1))
+        || (exponent >= exponent_m1)) {
+      return ret_t(-w);
+    }
+    ans = LOG_TWO + exponent_with_1mw - log1m_exp(exponent_with_1mw);
+    const auto diff_term = log1m_exp(exponent_with_w) - log1m_exp(exponent);
+    const auto log_w = log(w);
+    if (log_w > diff_term) {
+      ans += log_diff_exp(log_w, diff_term);
+      ans = exp(ans);
+    } else {
+      ans += log_diff_exp(diff_term, log_w);
+      ans = -exp(ans);
+    }
+  } else if (v > 0) {
+    const auto exponent_with_1mw = -2.0 * v * a * (1.0 - w);
+    const auto exponent = (-2 * a * v);
+    if ((exponent_with_1mw >= exponent_m1) || (exponent >= exponent_m1)) {
+      return ret_t(-w);
+    }
+    ans = LOG_TWO - log1m_exp(exponent_with_1mw);
+    const auto diff_term
+        = log_diff_exp(exponent_with_1mw, exponent) - log1m_exp(exponent);
+    const auto log_w = log(w);
+    if (log_w > diff_term) {
+      ans += log_diff_exp(log_w, diff_term);
+      ans = -exp(ans);
+    } else {
+      ans += log_diff_exp(diff_term, log_w);
+      ans = exp(ans);
+    }
+  } else {
+    return ret_t(-w);
+  }
+  if (fabs(ans) < INFTY) {
+    return ans;
+  } else {
+    return ret_t(NEGATIVE_INFTY);
+  }
+  return ans;
```

Comment:
I'm trying to parse the if statements here and think I made this a little more readable

```suggestion
  const auto v = -v_value;
  const auto w = 1 - w_value;
  int sign_v = v < 0 ? 1 : -1;
  const auto exponent_with_1mw = sign_v * 2.0 * v * a * (1.0 - w);
  const auto exponent = (sign_v * 2 * a * v);
  const auto exponent_with_w = 2 * a * v * w;
  if (unlikely(((exponent_with_1mw >= exponent_m1) || (exponent_with_w >= exponent_m1))
      || (exponent >= exponent_m1) || v == 0)) {
    return ret_t(-w);
  }
  ret_t diff_term;
  const auto log_w = log(w);
  if (v < 0) {
    ans = LOG_TWO + exponent_with_1mw - log1m_exp(exponent_with_1mw);
    diff_term = log1m_exp(exponent_with_w) - log1m_exp(exponent);
  } else if (v > 0) {
    ans = LOG_TWO - log1m_exp(exponent_with_1mw);
    diff_term
        = log_diff_exp(exponent_with_1mw, exponent) - log1m_exp(exponent);
  }
  if (log_w > diff_term) {
    ans += log_diff_exp(log_w, diff_term);
    ans = exp(ans);
    ans = sign_v == 1 ? ans : -ans;
  } else {
    ans += log_diff_exp(diff_term, log_w);
    ans = exp(ans);
    ans = sign_v == 1 ? -ans : ans;
  }
  if (unlikely(!is_scal_finite(ans))) {
    return ret_t(NEGATIVE_INFTY);
  }
  return ans;
```

Reply:
Ok. Comes with next commit.